src/app/icd/client/CheckInHandler.cpp

```diff
@@ -0,0 +1,176 @@
+#include <app/InteractionModelEngine.h>
+#include <app/OperationalSessionSetup.h>
+#include <protocols/secure_channel/Constants.h>
+
+namespace chip {
+namespace app {
+
+inline constexpr uint64_t kCheckInCounterMax = (1ULL << 32);
+inline constexpr uint32_t kKeyRefreshLimit   = (1U << 31);
+
+CheckInHandler::CheckInHandler() :
+    mOnConnectedCallback(HandleDeviceConnected, this), mOnConnectionFailureCallback(HandleDeviceConnectionFailure, this)
+{}
+
+CHIP_ERROR CheckInHandler::Init(Messaging::ExchangeManager * exchangeManager, ICDClientStorage * clientStorage,
+                                CheckInDelegate * delegate)
+{
+    VerifyOrReturnError(exchangeManager != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(clientStorage != nullptr, CHIP_ERROR_INVALID_ARGUMENT);
+    VerifyOrReturnError(mpExchangeManager == nullptr, CHIP_ERROR_INCORRECT_STATE);
+    VerifyOrReturnError(mpICDClientStorage == nullptr, CHIP_ERROR_INCORRECT_STATE);
+
+    mpExchangeManager  = exchangeManager;
+    mpICDClientStorage = clientStorage;
+    mpCheckInDelegate  = delegate;
+
+    return mpExchangeManager->RegisterUnsolicitedMessageHandlerForType(Protocols::SecureChannel::MsgType::ICD_CheckIn, this);
+}
+
+void CheckInHandler::Shutdown()
+{
+    mpICDClientStorage = nullptr;
+    mpCheckInDelegate  = nullptr;
+    if (mpExchangeManager)
+    {
+        mpExchangeManager->UnregisterUnsolicitedMessageHandlerForType(Protocols::SecureChannel::MsgType::ICD_CheckIn);
+        mpExchangeManager = nullptr;
+    }
+}
+
+CHIP_ERROR CheckInHandler::OnUnsolicitedMessageReceived(const PayloadHeader & payloadHeader, ExchangeDelegate *& newDelegate)
+{
+    // Return error for wrong message type
+    VerifyOrReturnError(payloadHeader.HasMessageType(Protocols::SecureChannel::MsgType::ICD_CheckIn),
+                        CHIP_ERROR_INVALID_MESSAGE_TYPE);
+
+    newDelegate = this;
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR CheckInHandler::OnMessageReceived(Messaging::ExchangeContext * ec, const PayloadHeader & payloadHeader,
+                                             System::PacketBufferHandle && payload)
+{
+    // If the message type is not ICD_CheckIn, return CHIP_NO_ERROR and exit
+    VerifyOrReturnError(payloadHeader.HasMessageType(Protocols::SecureChannel::MsgType::ICD_CheckIn), CHIP_NO_ERROR);
+
+    ByteSpan payloadByteSpan{ payload->Start(), payload->DataLength() };
+    ICDClientInfo clientInfo;
+    CounterType counter = 0;
+    // If the check-in message processing fails, return CHIP_NO_ERROR and exit.
+    VerifyOrReturnError(CHIP_NO_ERROR == mpICDClientStorage->ProcessCheckInPayload(payloadByteSpan, clientInfo, counter),
+                        CHIP_NO_ERROR);
+    CounterType receivedCheckInCounterOffset = (counter - clientInfo.start_icd_counter) % kCheckInCounterMax;
+
+    // Detect duplicate check-in messages and return CHIP_NO_ERROR on receiving a duplicate message
+    if (receivedCheckInCounterOffset <= clientInfo.offset)
```

Comment:
@mkardous-silabs @jepenven-silabs  it seems the counter first check-in message is same as the starting counter, then first check-in message would be dropped during testing, in code, i have not yet found the place that would do initial increment between commissioning complete and fist check-in message delivery. Either we allow the counter from first check-in message is same as the starting counter, or we update the server code to bump it. thanks

Reply:
@yunhanw-google can you create an issue and assign to me and jean-francois?

Reply:
https://github.com/project-chip/connectedhomeip/issues/31297, thanks